packages/marketplace/contracts/libraries/LibOrder.sol

```diff
@@ -33,37 +42,82 @@ library LibOrder {
+                        LibAsset.hash(order.takeAsset[0].assetType),
+                        order.salt
+                    )
+                );
+        } else {
+            bytes memory makeAssetsTypeEncoded;
+            for (uint i = 0; i < order.makeAsset.length; i++) {
+                makeAssetsTypeEncoded = abi.encodePacked(makeAssetsTypeEncoded, LibAsset.hash(order.makeAsset[i].assetType));
+            }
+
+            bytes memory takeAssetsTypeEncoded;
+            for (uint i = 0; i < order.takeAsset.length; i++) {
+                takeAssetsTypeEncoded = abi.encodePacked(takeAssetsTypeEncoded, LibAsset.hash(order.takeAsset[i].assetType));
+            }
+
+            return keccak256(
+                    abi.encode(
+                        order.maker,
+                        keccak256(makeAssetsTypeEncoded),
+                        keccak256(takeAssetsTypeEncoded),
+                        order.salt
+                    )
+                );
+        }
     }
 
     /// @notice Computes the complete hash of an order, including domain-specific data.
     /// @param order The order data.
+    /// @param version Order struct version - V1 or V2
     /// @return The complete hash of the order.
-    function hash(Order calldata order) internal pure returns (bytes32) {
-        return
+    function hash(Order calldata order, OrderType version) internal pure returns (bytes32) {
+        if(version == OrderType.V1){
+            return
             keccak256(
                 // solhint-disable-next-line func-named-parameters
                 abi.encode(
-                    ORDER_TYPEHASH,
+                    ORDER_TYPEHASH_V1,
                     order.maker,
-                    LibAsset.hash(order.makeAsset),
+                    LibAsset.hash(order.makeAsset[0]),
                     order.taker,
-                    LibAsset.hash(order.takeAsset),
+                    LibAsset.hash(order.takeAsset[0]),
                     order.salt,
                     order.start,
                     order.end
                 )
             );
+        } else {
+            bytes memory makeAssetsEncoded;
+            for (uint i = 0; i < order.makeAsset.length; i++) {
+                makeAssetsEncoded = abi.encodePacked(makeAssetsEncoded, LibAsset.hash(order.makeAsset[i]));
+            }
+
+            bytes memory takeAssetsEncoded;
+            for (uint i = 0; i < order.takeAsset.length; i++) {
+                takeAssetsEncoded = abi.encodePacked(takeAssetsEncoded, LibAsset.hash(order.takeAsset[i]));
+            }
+
+            return keccak256(abi.encode(
+                ORDER_TYPEHASH_V2,
+                order.maker,
+                keccak256(makeAssetsEncoded), // Hash of all makeAssets
+                order.taker,
+                keccak256(takeAssetsEncoded), // Hash of all takeAssets
+                order.salt,
+                order.start,
+                order.end
+            ));
+        }
     }
```

Comment:
would be great to find a way to make this versioning a bit more elegant than a bunch of `if else`. Could we use OOP to solve that ?

random example, just to get the idea:
```
LibOrderFactory.create(signature.version).hash(Order calldata order)
```
That wouldn't work because LibOrder is a library tho.